File: frontend/src/api/pages.ts
import {
  useQuery,
  useMutation,
  useQueryClient,
} from "@tanstack/react-query";
import {
  fetchJson,
  postJson,
  putJson,
  patchJson,
  deleteJson,
} from "@/api/client";
import type {
  PageSummary,
  PageDetail,
  PageCreateRequest,
  PageUpdateRequest,
  PageReorderRequest,
  TemplateSummary,
} from "@/types/pages";

const PAGES_KEY = ["pages"] as const;
const TEMPLATES_KEY = ["pages", "templates"] as const;

function pageDetailKey(slug: string) {
  return ["pages", slug] as const;
}

// ── Queries ─────────────────────────────────────────────────────────────────

/** GET /api/pages — sidebar listing of all user pages. */
export function usePages() {
  return useQuery({
    queryKey: PAGES_KEY,
    queryFn: () => fetchJson<PageSummary[]>("/api/pages"),
    staleTime: 30_000,
  });
}

/** GET /api/pages/:slug — single page with full layout. */
export function usePage(slug: string) {
  return useQuery({
    queryKey: pageDetailKey(slug),
    queryFn: () => fetchJson<PageDetail>(`/api/pages/${slug}`),
    staleTime: 30_000,
  });
}

/** GET /api/pages/templates — available templates for the create flow. */
export function usePageTemplates() {
  return useQuery({
    queryKey: TEMPLATES_KEY,
    queryFn: () => fetchJson<TemplateSummary[]>("/api/pages/templates"),
    staleTime: 10 * 60_000,
  });
}

// ── Mutations ───────────────────────────────────────────────────────────────

/** POST /api/pages — create a new page (blank or from template). */
export function useCreatePage() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: PageCreateRequest) =>
      postJson<PageDetail>("/api/pages", body),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: PAGES_KEY });
    },
  });
}

/** PUT /api/pages/:slug — update page name, icon, or layout. */
export function useUpdatePage(slug: string) {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: PageUpdateRequest) =>
      putJson<PageDetail>(`/api/pages/${slug}`, body),
    onSuccess: (updated) => {
      qc.setQueryData(pageDetailKey(slug), updated);
      qc.invalidateQueries({ queryKey: PAGES_KEY });
    },
  });
}

/** PATCH /api/pages/reorder — set page order. */
export function useReorderPages() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: PageReorderRequest) =>
      patchJson("/api/pages/reorder", body),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: PAGES_KEY });
    },
  });
}

/** DELETE /api/pages/:slug — delete a page. */
export function useDeletePage() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (slug: string) => deleteJson(`/api/pages/${slug}`),
    onSuccess: (_data, slug) => {
      qc.removeQueries({ queryKey: pageDetailKey(slug) });
      qc.invalidateQueries({ queryKey: PAGES_KEY });
    },
  });
}

File: analytics-api/app/routers/pages.py
"""Pages API — CRUD for user dashboard pages."""
import logging

from fastapi import APIRouter, Depends, HTTPException

from app.auth.dependencies import get_org_context
from app.models.auth import OrgContext
from app.models.pages import (
    PageCreateRequest,
    PageDetail,
    PageReorderRequest,
    PageSummary,
    PageUpdateRequest,
    TemplateSummary,
)
from app.services import page_service
from app.services.page_templates import TEMPLATES, TEMPLATE_LIST

logger = logging.getLogger(__name__)
router = APIRouter()


@router.get("/api/pages", response_model=list[PageSummary])
async def list_pages(ctx: OrgContext = Depends(get_org_context)):
    """List all pages for the current user (sidebar)."""
    return await page_service.get_user_pages(ctx.user_id)


@router.get("/api/pages/templates", response_model=list[TemplateSummary])
async def list_templates(ctx: OrgContext = Depends(get_org_context)):
    """List available page templates."""
    return [
        TemplateSummary(id=t.id, name=t.name, icon=t.icon, description=t.description)
        for t in TEMPLATE_LIST
    ]


@router.post("/api/pages", response_model=PageDetail, status_code=201)
async def create_page(
    body: PageCreateRequest,
    ctx: OrgContext = Depends(get_org_context),
):
    """Create a new page (blank or from a template)."""
    layout: list = []
    if body.template:
        t = TEMPLATES.get(body.template)
        if t is not None:
            layout = t.layout

    page = await page_service.create_page(
        user_id=ctx.user_id,
        org_id=ctx.org_id,
        name=body.name,
        icon=body.icon,
        layout=layout,
    )
    return page


@router.get("/api/pages/{slug}", response_model=PageDetail)
async def get_page(slug: str, ctx: OrgContext = Depends(get_org_context)):
    """Get a single page with its layout."""
    page = await page_service.get_page_by_slug(ctx.user_id, slug)
    if page is None:
        raise HTTPException(status_code=404, detail="Page not found")
    return page


@router.put("/api/pages/{slug}", response_model=PageDetail)
async def update_page(
    slug: str,
    body: PageUpdateRequest,
    ctx: OrgContext = Depends(get_org_context),
):
    """Update a page (name, icon, layout — all optional)."""
    page = await page_service.update_page(
        user_id=ctx.user_id,
        slug=slug,
        name=body.name,
        icon=body.icon,
        layout=body.layout,
    )
    if page is None:
        raise HTTPException(status_code=404, detail="Page not found")
    return page


@router.patch("/api/pages/reorder", status_code=204)
async def reorder_pages(
    body: PageReorderRequest,
    ctx: OrgContext = Depends(get_org_context),
):
    """Reorder pages by providing the full list of page_ids in desired order."""
    await page_service.reorder_pages(ctx.user_id, body.page_ids)


@router.delete("/api/pages/{slug}", status_code=204)
async def delete_page(slug: str, ctx: OrgContext = Depends(get_org_context)):
    """Delete a page."""
    deleted = await page_service.delete_page(ctx.user_id, slug)
    if not deleted:
        raise HTTPException(status_code=404, detail="Page not found")
